feat(api): Add module calibration mechanism to find square center on arbitrary position.

vegano1 opened the pull request:
# Overview

The geometries of modules vary from unit to unit, and even though when installed they are fastened to the deck there are variations that cause inaccuracies when performing actions around the deck. This is why we need to calibrate the position of the modules relative to the deck, to ensure accuracy, especially when moving labware around with the gripper.

Closes [RCORE-609](https://opentrons.atlassian.net/browse/RCORE-609)

This pull request starts to add the mechanism required to start a module calibration on the deck, where

* We find the nominal square of that slot
* We apply the offset from the module calibration definition given a specific module, which will apply the x, y, and z offset of the calibration square, NOTE: actual offsets will be added once we have the definitions.
* We use one of our `find_square_center` methods to find the center of the given module calibration square.


# Test Plan

- [x] Make sure we can still calibrate instruments
- [x] Make sure we can still calibrate the deck
- [x] Make sure we can call `calibrate_module` from ot3repl
- [x] Make sure we can start module calibration on any slot on the deck

# Changelog

- Separate loading of the nominal position of the calibration square given a slot from finding the actual position of that square on the deck. 
- This lets us find the calibration square center at an arbitrary position which is required for module calibration since the modules have different shapes and sizes and their calibration square might not be in-line with the deck slot calibration square.
- Able to start module calibration from ot3repl

# Review requests

- We don't yet have the module calibration labware, but this will be added in a future ticket
- Does this make sense?
- Is there anything I'm missing?


# Risk assessment

Low, Flex work


[RCORE-609]: https://opentrons.atlassian.net/browse/RCORE-609?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

vegano1 committed:
feat(api): add module calibration mechanism.

sfoster1 commented on api/src/opentrons/hardware_control/ot3_calibration.py:
```diff
@@ -740,6 +748,46 @@ async def _calibrate_mount(
+    # Find the estimated deck height. This will be used to baseline the edge detection points.
+    z_height = await find_deck_height(hcapi, mount, nominal_center)
+    nominal_center = nominal_center._replace(z=z_height)
+    LOG.info(f"Found deck at {z_height}mm")
```
nit but this is "module plate height" rather than "deck" right?

sfoster1 commented on api/src/opentrons/hardware_control/ot3_calibration.py:
```diff
@@ -821,3 +869,42 @@ async def calibrate_pipette(
+    """
+
+    try:
+        await hcapi.reset_instrument_offset(mount)
```
we do this when calibrating instruments because otherwise the offset we find will be an offset from the old offset, instead of an offset from 0 (and, since we're recalibrating the instrument, the old offset is probably wrong).

Here, we want this instrument offset to be the one that we've calibrated. Because of the way we implement instrument offsets, they apply to all logical deck coordinates: if you have a target deck position 200 that normally is machine position 200, and your offset is +10mm, then you'll move to machine position 210 instead of 200 when moving to the deck position. And if you're moving to deck 300, then you'll move to machine 310 just the same: offset +10 means that all motions end up +10, it's an offset.

Now let's bring in modules. Say the nominal position of a module is X=200. Say that its real position is X=210, so we should store that the module's calibration offset is +10. And say that deck=>machine is the identity + a 10mm offset (so deck 200 = machine 210 ideally).

Now say that you have an instrument offset of 0, as you do after this reset. When you go to start calibration, you look up the module nominal and you get deck 200. Since your offset is 0, you turn that into machine 200, and move there. The thing is, in reality you're now at deck 190, since you haven't added those 10mm.

At deck 190, you'll run calibration and say, "ah, this module is actually +20 of where I expected it", and save the a calibration of +20, because the instrument offset applies to everything.

Finally, after restoring the old offset, you might load up the calculated module offset, and go 10mm too far.

All in all, don't reset the instrument offset here.

vegano1 replied:
This is actually critical, thanks for pointing this out.

sfoster1 reviewed (changes requested):
Structure looks good and I like the refactor but definitely don't reset the instrument offset

vegano1 committed:
fix comment
DO NOT reset previous instrument offset when calibrating module

sfoster1 commented on api/src/opentrons/hardware_control/ot3_calibration.py:
```diff
@@ -740,6 +748,46 @@ async def _calibrate_mount(
+
+    # Find the estimated module plate height. This will be used to baseline the edge detection points.
+    z_height = await find_deck_height(hcapi, mount, nominal_center)
+    nominal_center = nominal_center._replace(z=z_height)
```
we should make this `test_center = nominal_center._replace(z=z_height)` rather than altering `nominal_center` because...

sfoster1 commented on api/src/opentrons/hardware_control/ot3_calibration.py:
```diff
@@ -740,6 +748,46 @@ async def _calibrate_mount(
+
+    # Find the module square center using the given method
+    found_center = await find_slot_center(hcapi, mount, nominal_center, method)
+    offset = nominal_center - found_center
```
since `nominal_center` has its z coordinate replaced with the detected z position of the plate, the z coordinate here will always be 0. if instead we make `test_center = nominal_center._replace(z=z_height)` and then `found_center = await find_slot_center(..., test_center, ...)` we won't have this problem

sfoster1 reviewed (changes requested):
A couple things - first, there's a pretty straightforward issue with the way the z height is stored, but second, looking at the way we had to implement the position lookup here is making me a little uncomfortable. The hardware control API doesn't really have any concept of geometry systems, sort of on purpose; this feels like a bad place to do the stackup. Having it know the deck layout is kind of on the edge, though it is bolted on so maybe it's ok; but having it know about module _geometries_ like this, as opposed to module behavior, feels like too much.

What if we do one more stage of refactor in the following way:
- Make a method called like `find_calibration_structure_position(hcapi, mount, nominal_center) -> Point` that calls `find_deck_height` and `find_slot_center` (maybe rename that to `find_calibration_structure_height` and `find_calibration_structure_center`
- replace basically all the body of `_calibrate_module` and `_calibrate_mount` with a call to this method
- hoist the position determination stuff as high as it will go, ideally to the top-level `calibrate_module` call (i guess we don't have to change it for the mounts yet)
This puts us in a pretty good position for the next PR that integrates this more fully to actually do the geometry calculation somewhere else: in the engine command that drives module calibration. At that time, we change the signature of `calibrate_module` to be `calibrate_module(hcapi, mount, nominal_center)`; we get the nominal center in the engine from loading a module geometry and loading a calibration plate onto the module and asking for its center.

Ideally we'll also do a refactor for a similar thing in the mount calibration commands since they also use the engine, but that can be later, and we'll be out of the business of geometry calculations in the hardware controller again.

vegano1 committed:
prep calibrate_module mechanism so we arent relying on geometric calculations in the hardware controller as that belongs in the protocol engine

vegano1 committed:
fix linting

sfoster1 reviewed (approved):
Looks awesome! Nice work.

vegano1 committed:
Merge branch 'edge' into RCORE-609_autocalibration_on_arbitrary_position

Laura-Danielle commented on api/src/opentrons/hardware_control/ot3_calibration.py:
```diff
@@ -250,7 +252,7 @@ def _get_calibration_square_position_in_slot(slot: int) -> Point:
 
 
-async def find_deck_height(
+async def find_calibration_structure_height(
```
Can we change the doc string description here if we're going to be using this to find other things besides deck height?

Laura-Danielle replied:
We should change the error too inside this function.

Laura-Danielle commented on api/src/opentrons/hardware_control/ot3_calibration.py:
```diff
@@ -673,6 +675,26 @@ async def find_slot_center_noncontact(
+) -> Point:
+
+    # Perform xy offset search
+    if method == CalibrationMethod.LINEAR_SEARCH:
```
We're removing `LINEAR_SEARCH` (see [alise's pr](https://github.com/Opentrons/opentrons/pull/12243#pullrequestreview-1350559702)) so we should remove this option and default to BINARY.

ahiuchingau replied:
I agree with Laura here - I think we should just remove this right now and have the default method be `BINARY_SEARCH`

Laura-Danielle commented on api/src/opentrons/hardware_control/ot3_calibration.py:
```diff
@@ -740,6 +744,42 @@ async def _calibrate_mount(
+    """Find the calibration square offset given an arbitry postition on the deck."""
+    # Find the estimated structure plate height. This will be used to baseline the edge detection points.
+    z_height = await find_calibration_structure_height(hcapi, mount, nominal_center)
+    test_center = nominal_center._replace(z=z_height)
```
can we maybe call this `initial_center` or something?

Laura-Danielle commented on api/src/opentrons/hardware_control/ot3_calibration.py:
```diff
@@ -821,3 +861,43 @@ async def calibrate_pipette(
+    or the probe has been lowered). We also need to have the module
+    prepped for calibration (for example, not heating, lid closed,
+    not shaking, etc) and the corresponding module calibration
+    block placed in the module slot. The robot should be homed.
```
```suggestion
    block placed in the module slot.
    
    The robot should be homed before calling this function.
```

Laura-Danielle commented on api/src/opentrons/hardware_control/ot3_calibration.py:
```diff
@@ -821,3 +861,43 @@ async def calibrate_pipette(
+
+    try:
+        # add the probe depending on the mount
+        if mount == OT3Mount.GRIPPER:
```
Wait, are we going to need to use the gripper to calculate module offsets? Why can't we restrict to pipettes only to get the module offset? (cc @sfoster1 )

vegano1 replied:
I agree that a user is most likely to have a pipette than a gripper, but this is mostly for flexibility.

sfoster1 replied:
Yeah I guess the answer is "there's no reason for the restriction so why add it"

Laura-Danielle replied:
yeah I guess I'm just thinking from a user-flow perspective. Since we're going to be most likely restricting deck calibration to pipettes, I think we should do the same to module calibration until told otherwise.

Laura-Danielle reviewed (changes requested):
Mostly just some quick copy clean-up and one larger question about whether we should restrict calibrating modules with pipettes only.

vegano1 committed:
clean up from code review

vegano1 committed:
Merge branch 'edge' into RCORE-609_autocalibration_on_arbitrary_position

vegano1 committed:
comment fixup

vegano1 committed:
fix hardware-testing script

vegano1 committed:
fix test script

Laura-Danielle commented on api/src/opentrons/hardware_control/ot3_calibration.py:
```diff
@@ -66,10 +68,10 @@ class CalibrationMethod(Enum):
 
-class DeckNotFoundError(RuntimeError):
-    def __init__(self, deck_height: float, lower_limit: float) -> None:
+class StructureNotFoundError(RuntimeError):
```
```suggestion
class CalibrationStructureNotFoundError(RuntimeError):
```

Laura-Danielle reviewed (approved):
Changes look good, but make sure to coordinate with Alise before merging this.

vegano1 committed:
rename structure not found error so its more descriptive
we are removing LINEAR_SEARCH  method in a future pr, so change the function signatures so we default to BINARY_SEARCH so its easier to merge.
